File: algorithm/src/leetcode/IntegerToRoman_12.java
package leetcode;

// #12 - Integer To Roman
public class IntegerToRoman_12 {
	public static void main(String[] args) {
		int num = 3;
		System.out.println("intToRoman = " + intToRoman(num));
	}

	public static String intToRoman(int num) {
		int[] values = {1000,900,500,400,100,90,50,40,10,9,5,4,1};
		String[] romanLetters = {"M","CM","D","CD","C","XC","L","XL","X","IX","V","IV","I"};

		StringBuilder answer = new StringBuilder();
		for(int i=0;i<values.length;i++) {
			while(num >= values[i]) {
				num = num - values[i];
				answer.append(romanLetters[i]);
			}
		}
		return answer.toString();
	}
}

File: algorithm/src/leetcode/ruby/IntegerToRoman_12.rb
# #12 Int to Roman

# @param {Integer} num
# @return {String}
def int_to_roman(num)
  data = { 1 => "I", 4 => "IV", 5 => "V", 9 => "IX", 10 => "X", 40 => "XL",
           50 => "L", 90 => "XC", 100 => "C", 400 => "CD", 500 => "D", 900 => "CM", 1000 => "M" }

  output = ""
  return data[num] if data.key?(num)

  data.keys.reverse.each do |value|
    while (num >= value)
      return output + "#{data[num]}" if data.key?(num)

      num -= value
      map_val = data[value]
      output += "#{map_val}"
    end
  end

  output
end

num = 3
p int_to_roman(num)